xapian-core/common/serialise-double.cc

```diff
@@ -38,175 +40,159 @@
+		  "Check if size of double and 64 bit int is same");
+    memcpy(&temp, &v, sizeof(double));
+    temp = do_bswap(temp);
+    return string(reinterpret_cast<const char *>(&temp), sizeof(double));
+# else
+    return string(reinterpret_cast<const char *>(&v), sizeof(double));
+# endif
 }
 
-std::string serialise_double(double v)
+double unserialise_double(const char ** p, const char * end)
 {
-    /* First byte:
-     *  bit 7 Negative flag
-     *  bit 4..6 Mantissa length - 1
-     *  bit 0..3 --- 0-13 -> Exponent + 7
-     *            \- 14 -> Exponent given by next byte
-     *             - 15 -> Exponent given by next 2 bytes
+    if (end - *p < 8) {
+	throw Xapian::SerialisationError(
+	    "Bad encoded double: insufficient data");
+    }
+    double result;
+# ifdef WORDS_BIGENDIAN
+    uint64_t temp;
+    static_assert(sizeof(temp) == sizeof(double),
+		  "Check if size of double and 64 bit int is same");
+    memcpy(&temp, *p, sizeof(double));
+    temp = do_bswap(temp);
+    memcpy(&result, &temp, sizeof(double));
+# else
+    memcpy(&result, *p, sizeof(double));
+# endif
+    *p += 8;
+    return result;
+}
+
+#else
+
+string serialise_double(double v){
+    /*
+     * First bit(msb) -> sign (1 means negative)
+     * next 11 bits -> exponent
+     * last 52 bits -> mantissa
      *
-     * Then optional medium (1 byte) or large exponent (2 bytes, lsb first)
+     * frexp gives fraction within the range [0.5, 1)
+     * We multiply it by 2 to change the range to [1.0, 2.0)
+     * and reduce exp by 1, since this is the way doubles
+     * are stored in IEEE-754.
      *
-     * Then mantissa (0 iff value is 0)
+     * Conversion of mantissa to bits is done by
+     * multiplying the mantissa with 2^52, converting
+     * it to a 64 bit integer representation of the original
+     * double.
      */
 
-    bool negative = (v < 0.0);
+    static_assert((uint64_t)1 << 52 <= numeric_limits<double>::max(),
+		  "Check if 2^52 can be represented by a double");
 
-    if (negative) v = -v;
-
-    int exp = base256ify_double(v);
-
-    string result;
-
-    if (exp <= 6 && exp >= -7) {
-	unsigned char b = static_cast<unsigned char>(exp + 7);
-	if (negative) b |= static_cast<unsigned char>(0x80);
-	result += char(b);
-    } else {
-	if (exp >= -128 && exp < 127) {
-	    result += negative ? char(0x8e) : char(0x0e);
-	    result += char(exp + 128);
-	} else {
-	    if (exp < -32768 || exp > 32767) {
-		throw Xapian::InternalError("Insane exponent in floating point number");
-	    }
-	    result += negative ? char(0x8f) : char(0x0f);
-	    result += char(unsigned(exp + 32768) & 0xff);
-	    result += char(unsigned(exp + 32768) >> 8);
-	}
+    uint64_t result = 0;
+
+    if (v == 0.0) {
+	result = 0.0;
+	return string(reinterpret_cast<const char *>(&result),
+		      sizeof(uint64_t));
     }
 
-    int maxbytes = min(MAX_MANTISSA_BYTES, 8);
+    bool negative = (v < 0.0);
+    if (negative) {
+	v = -v;
+	result |= (uint64_t)1 << 63;
```

Comment:
The coding guidelines in `HACKING` say to avoid C-style casts:

```
* Use C++ style casts (static_cast<>, reinterpret_cast<>, and const_cast<>)  
  or constructor-syntax (e.g. ``double(value)``) in preference to C style
  casts.  The syntax of the C++ casts is ugly, but they do make the
  intent much clearer which is definitely a good thing, and they avoid issues
  such as casting away const when you only meant to cast the type of a pointer.
```

So here `uint64_t(1)` (or `static_cast<uint64_t>(1)` is preferrable (and similarly in a couple of other places).

We'd like to be able to enable the GCC warning `-Wold-style-cast` which would issue a warning for such cases:

> -Wold-style-cast (C++ and Objective-C++ only)
>
>    Warn if an old-style (C-style) cast to a non-void type is used within a C++ program. The new-style casts (dynamic_cast, static_cast, reinterpret_cast, and const_cast) are less vulnerable to unintended effects and much easier to search for.

https://gcc.gnu.org/onlinedocs/gcc/C_002b_002b-Dialect-Options.html

However, currently we've had to disable this warning due to it triggering for macros in system header files on some platforms, but at some point hopefully we can turn it on again.

Reply:
Fixed accordingly.